From 8c9537529bc10fdfe9520256fc42126764fb0efc Mon Sep 17 00:00:00 2001
From: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
Date: Mon, 8 Sep 2014 12:59:12 +0100
Subject: [PATCH 326/351] mfd: arizona: Disable DVFS boost in suspend

When going into suspend disable the DVFS boost since it
is illegal to have this enabled when DCVDD != 1.8v

Change-Id: I2cd83bcae79dfcfefd3aac2502c4ab9807f767da
Signed-off-by: Richard Fitzgerald <rf@opensource.wolfsonmicro.com>
---
 drivers/mfd/arizona-core.c |   40 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 40 insertions(+)

diff --git a/drivers/mfd/arizona-core.c b/drivers/mfd/arizona-core.c
index 693662c..828ff6b 100644
--- a/drivers/mfd/arizona-core.c
+++ b/drivers/mfd/arizona-core.c
@@ -614,6 +614,32 @@ static int arizona_runtime_resume(struct device *dev)
 		goto err;
 	}
 
+	switch(arizona->type) {
+	case WM5102:
+	case WM8997:
+	case WM8998:
+	case WM1814:
+		/* Restore DVFS setting */
+		ret = 0;
+		mutex_lock(&arizona->subsys_max_lock);
+		if (arizona->subsys_max_rq != 0) {
+			ret = regmap_update_bits(arizona->regmap,
+				ARIZONA_DYNAMIC_FREQUENCY_SCALING_1,
+				ARIZONA_SUBSYS_MAX_FREQ, 1);
+		}
+		mutex_unlock(&arizona->subsys_max_lock);
+
+		if (ret != 0) {
+			dev_err(arizona->dev,
+				"Failed to enable subsys max: %d\n",
+				ret);
+			goto err;
+		}
+		break;
+	default:
+		break;
+	}
+
 	return 0;
 
 err:
@@ -629,6 +655,20 @@ static int arizona_runtime_suspend(struct device *dev)
 
 	dev_dbg(arizona->dev, "Entering AoD mode\n");
 
+	switch(arizona->type) {
+	case WM5102:
+	case WM8997:
+	case WM8998:
+	case WM1814:
+		/* Must disable DVFS boost before powering down DCVDD */
+		regmap_update_bits(arizona->regmap,
+			ARIZONA_DYNAMIC_FREQUENCY_SCALING_1,
+			ARIZONA_SUBSYS_MAX_FREQ, 0);
+		break;
+	default:
+		break;
+	}
+
 	if (arizona->external_dcvdd) {
 		ret = regmap_update_bits(arizona->regmap,
 					 ARIZONA_ISOLATION_CONTROL,
-- 
1.7.9.5

